django/db/models/deletion.py

```diff
@@ -219,7 +220,12 @@ def collect(self, objs, source=None, nullable=False, collect_related=True,
                     if self.can_fast_delete(sub_objs, from_field=field):
                         self.fast_deletes.append(sub_objs)
                     elif sub_objs:
-                        field.remote_field.on_delete(self, field, sub_objs, self.using)
+                        try:
+                            field.remote_field.on_delete(self, field, sub_objs, self.using)
+                        except ProtectedError as e:
+                            protected_objects += e.protected_objects
+            if protected_objects:
+                raise ProtectedError('', protected_objects)
```

Comment:
You might have thought of this already, but it might be helpful to try to pass along the first error message, though it wouldn't be quite as clean, and it gets caught anyway.